File: src/main/java/io/github/jsnimda/common/config/CategorizedConfigOptions.java
package io.github.jsnimda.common.config;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.commons.lang3.tuple.Pair;

public class CategorizedConfigOptions extends ConfigOptionBase implements IConfigElementResettableMultiple {

  private List<Pair<String, List<IConfigOption>>> categories = new ArrayList<>();
  private List<IConfigOption> currentCategory = null;
  private Map<String, IConfigOption> configOptionsMap = new HashMap<>();

  public void addCategory(String categoryName) {
    currentCategory = new ArrayList<>();
    categories.add(Pair.of(categoryName, currentCategory));
  }

  public void addConfigOption(IConfigOption configOption) {
    if (currentCategory == null) {
      addCategory("");
    }
    currentCategory.add(configOption);
    configOptionsMap.put(configOption.getKey(), configOption);
  }

  public List<Pair<String, List<IConfigOption>>> getCategories() {
    return categories;
  }

  @Override
  public Map<String, ? extends IConfigOption> getConfigOptionsMap() {
    return configOptionsMap;
  }

}

File: src/main/java/io/github/jsnimda/common/config/CategorizedConfigOptions.kt
package io.github.jsnimda.common.config


class CategorizedConfigOptions : ConfigOptionBase(), IConfigElementResettableMultiple {
  val categories: LinkedHashMap<String, List<IConfigOption>> = linkedMapOf()
  private var _currentCategory: MutableList<IConfigOption>? = null
  private var currentCategory: MutableList<IConfigOption>
    get() = _currentCategory ?: addCategory("")
    set(value) {
      _currentCategory = value
    }

  fun addCategory(categoryName: String) = mutableListOf<IConfigOption>().also {
    currentCategory = it
    categories[categoryName] = it
  }

  fun addConfigOption(configOption: IConfigOption) {
    currentCategory.add(configOption)
  }

  override fun getConfigOptionsMap() = getConfigOptionsMapFromList()

  override fun getConfigOptionsList() = categories.values.flatten()
}